packages/protocol/src/HarvestManager.sol

```diff
@@ -0,0 +1,270 @@
+  error HarvestManager__harvest_vaultNotAllowed();
+
+  address public immutable treasury;
+
+  uint256 public constant MAX_FEE = 1e18; // 100%
+  uint256 public constant MIN_FEE = 1e16; // 1%
+  uint256 public constant BASE_FEE = 2e17; // 20%
+
+  uint256 public protocolFee;
+
+  /// @dev addresses allowed to harvest via this manager.
+  mapping(address => bool) public allowedExecutor;
+
+  address private currentVaultHarvest;
+
+  constructor(address chief_, address treasury_) {
+    __SystemAccessControl_init(chief_);
+    treasury = treasury_;
+    protocolFee = BASE_FEE;
+  }
+
+  /// @inheritdoc IHarvestManager
+  function allowExecutor(address executor, bool allowed) external override onlyTimelock {
+    if (executor == address(0)) {
+      revert HarvestManager__zeroAddress();
+    }
+    if (allowedExecutor[executor] == allowed) {
+      revert HarvestManager__allowExecutor_noAllowChange();
+    }
+    allowedExecutor[executor] = allowed;
+    emit AllowExecutor(executor, allowed);
+  }
+
+  /// @inheritdoc IHarvestManager
+  function setFee(uint256 fee) external override onlyTimelock {
+    if (fee == protocolFee) {
+      revert HarvestManager__setFee_noFeeChange();
+    }
+    if (fee > MAX_FEE) {
+      revert HarvestManager__setFee_feeTooHigh();
+    }
+    if (fee < MIN_FEE) {
+      revert HarvestManager__setFee_feeTooLow();
+    }
+    protocolFee = fee;
+    emit SetFee(fee);
+  }
+
+  /// @inheritdoc IHarvestManager
+  function harvest(
+    IVault vault,
+    Strategy strategy,
+    IHarvestable provider,
+    ISwapper swapper,
+    bytes memory data
+  )
+    external
+  {
+    if (!allowedExecutor[msg.sender]) {
+      revert HarvestManager__harvest_notValidExecutor();
+    }
+    if (
+      (strategy == Strategy.ConvertToCollateral || strategy == Strategy.RepayDebt)
+        && !chief.allowedSwapper(address(swapper))
+    ) {
+      revert HarvestManager__harvest_notValidSwapper();
+    }
+
+    if (currentVaultHarvest != address(0)) {
+      revert HarvestManager__harvest_harvestAlreadyInProgress();
+    }
+
+    currentVaultHarvest = address(vault);
+
+    vault.harvest(strategy, provider, swapper, data);
+  }
+
+  function completeHarvest(
+    address vault,
+    Strategy strategy,
+    IHarvestable provider,
+    ISwapper swapper,
+    address[] memory tokens,
+    uint256[] memory amounts
+  )
+    external
+    returns (bytes memory data)
+  {
+    if (msg.sender != currentVaultHarvest) {
+      revert HarvestManager__harvest_vaultNotAllowed();
+    }
+
+    IVault vault_ = IVault(vault);
+
+    if (strategy == Strategy.ConvertToCollateral) {
+      data = _convertToCollateral(tokens, amounts, swapper, vault_);
+    }
+    if (strategy == Strategy.RepayDebt) {
+      data = _repayDebt(provider, tokens, amounts, swapper, vault_);
+    }
+    if (strategy == Strategy.Distribute) {
+      data = _distribute(tokens, amounts, vault_);
+    }
+
+    currentVaultHarvest = address(0);
+  }
```

Comment:
Completing the harvest through the HarvestManager should also emit an Event. 
This should not overlap with the information on the event that should also be emitted on the vault itself. 
I probably suggest that this event here focuses on the treasury profit and what the vault "gained" by the strategy.